### ISSUE TITLE
Request: Add optional Checker to `enforce explicit namespace packages` aka `there must be an __init__.py in each subdir`

### ISSUE THREAD

**Issue by doublethefish** [PyCQA/pylint#7085]
### Current problem

Often tools (like Sphinx and pytest) do strange things because of `importlib`'s various quirks (see [importlib](https://docs.python.org/3/library/importlib.html) and [PEP420](https://peps.python.org/pep-0420/)*), specifically associated with `implicit namespaces` (vs explicit).

One explicit (and important) example is `coverage.py` + `pytest` doesn't always include files in implicit namespace packages - so code coverage is reported incorrectly 🤬. Obviously that's not the fault of `pylint` but having a pylint Checker to detect such problems would be ace.

Those quirks are most often solved if one enforces `__init__.py` files to be in every package subdir 🎊.

\* Sometimes, I wonder if PEP420 was conceived on the 20th April, sigh 😉.

### Desired solution

In my most nebulous projects, I have a `pytest` test that checks various things, but I am not sure if it is appropriate as a Checker as-is.

One day I might get around to integrating it into `pylint` but it would probably be a good first feature for someone to implement.

Here's some code that "does the trick".
Note: that it contains `no cover` declaratives only because we would be testing the test code, but `pylint`'s test infrastructure is very well set up to test this sort of code as a checker.

```py
# tests that all the conditions to get coverage are met
import os
import typing
from pathlib import Path


def _dir_has_any_pyfile(
    start_dir: str,
) -> typing.Union[bool, typing.List[str]]:  # pragma: no cover
    for _dir_name, _subdirs, file_list in os.walk(start_dir):
        subdir_pyfiles = [
            filename for filename in file_list if filename.endswith(".py")
        ]
        if subdir_pyfiles:
            return subdir_pyfiles

    # not found
    return False


def test_all_dirs_have_inits() -> None:
    dirs_with_no_init = []
    dirs_with_no_init_but_subdir_pys = []
    dirs_with_only_init = []

    base_path = (Path(os.path.dirname(__file__)) / ".." / "..").resolve()
    paths_to_test = (base_path / "mdgm", base_path / "tests")

    for start_dir in paths_to_test:
        for dir_name, subdirs, file_list in os.walk(start_dir):
            # strip out known util dirs
            subdirs[:] = [
                dname
                for dname in subdirs
                if dname not in ("__pycache__", ".pytest_cache")
            ]

            if str(Path(dir_name).relative_to(base_path)) == "tests/data":
                subdirs[:] = []
                continue

            pyfiles = [filename for filename in file_list if filename.endswith(".py")]
            subdir_pyfiles = None
            rel_to_root = str(Path(dir_name).relative_to(base_path))

            if not pyfiles:  # pragma: no cover
                subdir_pyfiles = _dir_has_any_pyfile(rel_to_root)
                if not subdir_pyfiles:
                    # we may or may not have a dir without python files
                    continue  # skip as no python files

            if "__init__.py" not in file_list:  # pragma: no cover
                if pyfiles:
                    dirs_with_no_init.append(rel_to_root)
                else:
                    assert subdir_pyfiles, (
                        "if we get here, one of pyfiles *or* "
                        "subdir_pyfiles should contain files as "
                        "we should have 'continued' otherwise"
                    )
                    dirs_with_no_init_but_subdir_pys.append(rel_to_root)
            else:
                single_file_no_dirs = (len(file_list) == 1) and (len(subdirs) == 0)
                if single_file_no_dirs:  # pragma: no cover
                    dirs_with_only_init.append(rel_to_root)

    assert not dirs_with_no_init, (
        "We have directories with missing '__init__.py' file\n"
        f"\t{dirs_with_no_init}\n"
    )

    assert not dirs_with_only_init, (
        "We have directories that _only_ contain an __init__.py file and"
        f"nothing else!\n\t{dirs_with_only_init}"
    )

    assert not dirs_with_no_init_but_subdir_pys, (
        "We have directories with missing '__init__.py' file but "
        ".py files in subdirs\n"
        f"\t{dirs_with_no_init_but_subdir_pys}\n"
    )

```

### Additional context

I have been meaning to create a checker for this for some time. I was inspired to create it after re-reading my post comment (https://github.com/PyCQA/pylint/issues/4437#issuecomment-961823805) on  #4437

**Comment by Pierre-Sassoulas** [comment]
I think this is a valuable check for code directory, but it would probably have a lot of false positives for tests directory. ``pytest`` is adding all the test directory to the python path and strongly suggest to not use ``__init__.py`` in its test directory. How would we detect the difference between the two ? Maybe the root directory named ``test`` or ``tests`` ?

**Comment by Pierre-Sassoulas** [comment]
I remembered having very hard to debug problems because of that. So I checked and I think [the first layout here](https://docs.pytest.org/en/latest/explanation/goodpractices.html#tests-outside-application-code) still asks for no ``__init__.py`` in test directory. But it's clearly possible because to use ``__init__.py`` everywhere also it also avoid name collision ! So it seems this check is actually a **very good** thing for pytest users.

**Comment by doublethefish** [comment]
> So it seems this check is actually a very good thing for pytest users.

👍

> I remembered having very hard to debug problems because of that.

Assume "that" refers to having a`tests/__init__.py` file; yes I have had problems with it too, one of the worst ones was in `pylint` (if you remember?).

**Comment by Pierre-Sassoulas** [comment]
>  (if you remember?).

I do remember 😄. Btw remembering it saved my colleagues weeks of debug at work, so I don't think it's a problem that never happen. But now that I read the documentation it seems we could have just added ``__init__.py`` everywhere to fix the issue ? Do you understand it that way too ?

**Comment by doublethefish** [comment]
> Do you understand it that way too ?

Yep, I think so. An additional benefit of the Checker I guess.
